### ISSUE TITLE
Is this the interpreter pattern?

### ISSUE THREAD

**Issue by AriaMinaei** [redux-effects/redux-effects#2]
I'm still wrapping my head around all the functional patterns. Reading redux-effects, it looks similar to the Free monad with an interpreter.

If it is, would it make sense for it to implement the fantasy-land spec so we can use it with the other familiar tools built around fantasy-land?

**Comment by ashaffer** [comment]
I'm not familiar with the interpreter pattern.  So, maybe?  Would you mind explaining or providing a link to read more about it?

EDIT: I mean, in a trivial sense it is equivalent to this: https://en.wikipedia.org/wiki/Interpreter_pattern  But I assume you mean something a bit more specific.

**Comment by AriaMinaei** [comment]
Yes, the wiki link you provided seems to be about the same thing. I found the pattern in Haskell literature. The goal seems to be similar to redux-effects` goal.

The basic idea is that you accumulate the intents of some action into some sort of container (e.g. `read cookie A`, then `read cookie B`, then `transform their value into some string`, then `send the string to the server`, and finally `return something based on the server's response`), and then hand the container to an interpreter to perform the side-effects.

The container can be a Free monad, which would enable us to declaratively write the intents of a function while using the methods available to all monads, so the code doesn't look too cryptic. [This SO answer](http://programmers.stackexchange.com/a/242803/16536) describes the pattern very well.

Anyway, I don't know how sensible it would be to re-implement the pattern in JS similar to how it's implemented in Haskell, but it's worth exploring. It might yield a clean and familiar syntax.

I should also mention [fantasy-frees](https://github.com/fantasyland/fantasy-frees) which is actively exploring this idea.

**Comment by ashaffer** [comment]
Ah, yes.  That is very similar indeed.  I considered using Haskell's monad `return/bind` terms, but I figured promises would probably be more familiar to JS developers.  The cool thing about this though is that you could just make your own action creators that use that syntax, since I think they are functionally equivalent.

I'd definitely be curious about making it fantasy-land compatible, though I wouldn't want to add too much bloat/complexity to be able to do that.

**Comment by AriaMinaei** [comment]
Great! I'll watch this repository to see how these experiments turn out. I'll be very interested to know how this might help in real-world projects.

I toyed with the idea of using interpreters as action creators in my current project, but I decided it wasn't worth the effort for this particular project. I don't need server-side rendering, and most of my action creators are too simple to warrant heavy testing. I've seen using ADTs is helping me make functions more "obviously correct" and that is saving me time on testing.

But looking at this repo, I realize they can be implemented more easily than I thought. 

Anyway, thanks for your answer, and thank you for doing this in the open :+1:
